### ISSUE TITLE
Realm: Assertion `datalen == (impl->redop->sizeof_lhs * (trigger_args.internal.trigger_gen - trigger_args.internal .previous_gen))' failed.

### ISSUE THREAD

**Issue by syamajala** [StanfordLegion/legion#1809]
I'm seeing the following assertion when running S3D on Frontier:
```
s3d.x: /lustre/orion/cmb103/scratch/seshuy/legion_s3d_tdb/legion/runtime/realm/barrier_impl.cc:1259: static void Realm:
:BarrierTriggerMessage::handle_message(Realm::NodeID, const Realm::BarrierTriggerMessage&, const void*, size_t, Realm::
TimeLimit): Assertion `datalen == (impl->redop->sizeof_lhs * (trigger_args.internal.trigger_gen - trigger_args.internal
.previous_gen))' failed.
```

It seems to happen when I turn on profiling as my run without profiling succeeded.

**Comment by syamajala** [comment]
Also I am running on the onepool branch.

**Comment by lightsighter** [comment]
The critical path profiling requires the use of Realm barrier reductions which are not used when you are not profiling. What are the values of `datalen`, `impl->redop->sizeof_lhs`, `trigger_args.internal.trigger_gen`, `trigger_args.internal.previous_gen`, and `trigger_args.internal.first_gen`?

**Comment by lightsighter** [comment]
If you want to avoid this error for now and you don't mind not having critical path profiling then you can run with `-lg:prof_no_critical_paths`.

I'll also note that this doesn't have anything to do with `onepool` and will happen in the master branch as well.

**Comment by apryakhin** [comment]
Yep that is certainly a bug due to my recent merge of the notification broadcast. That will be fixed at the earliest convenience.

**Comment by apryakhin** [comment]
That would be the patch:
- https://gitlab.com/StanfordLegion/legion/-/merge_requests/1613

That needs a test to be added which I will do shortly and then it's ready

**Comment by apryakhin** [comment]
@syamajala The assert you are hitting is only on the `onepool` branch? I wonder if we can actually test the path with S3D

**Comment by lightsighter** [comment]
The onepool branch is merged into the master branch now, so you can upstream those changes to whatever feature branch you are testing.

**Comment by apryakhin** [comment]
PR1613 should already include changes from `one pool` branch then. @syamajala just to follow up again on testing the fix

**Comment by syamajala** [comment]
Will test the pull request.

**Comment by syamajala** [comment]
Frontier is down today.

**Comment by syamajala** [comment]
Job is in the queue, but it looks like the compute nodes are still down on Frontier.

**Comment by syamajala** [comment]
I tested on sapling and it seems to be working.

**Comment by syamajala** [comment]
oops. meant to click comment not close. Please close after you merge to master.

**Comment by apryakhin** [comment]
@syamajala Can you pull the changes from master and see if that works? We made a fix already which I think should handle it:
- https://gitlab.com/StanfordLegion/legion/-/merge_requests/1630

**Comment by apryakhin** [comment]
@syamajala When you get a chance if you could please follow up on this and if the issue not re-occurring please consider closing it.

**Comment by syamajala** [comment]
It works.

**Comment by syamajala** [comment]
I think Im hitting this again?

```
python: /tmp/conda-croot/legate/work/arch-conda/skbuild_core/_deps/legion-src/runtime/realm/barrier_impl.cc:1365: void 
Realm::BarrierImpl::handle_remote_trigger(Realm::NodeID, Realm::ID::IDType, Realm::EventImpl::gen_t, Realm::EventImpl::
gen_t, Realm::EventImpl::gen_t, Realm::ReductionOpID, Realm::NodeID, int, std::vector<Realm::RemoteNotification>, int, 
bool, unsigned int, const void*, size_t, Realm::TimeLimit): Assertion `datalen == (redop->sizeof_lhs * (original_gen - 
previous_gen))' failed.
```

**Comment by syamajala** [comment]
Yes. I believe Im hitting this again. My run worked when I ran with `-lg:prof_no_critical_paths`

**Comment by syamajala** [comment]
@apryakhin this issue is still there for s3d

**Comment by apryakhin** [comment]
@syamajala That must have been fixed with https://gitlab.com/StanfordLegion/legion/-/merge_requests/1766. Can you confirm whether you still see the issue?

**Comment by syamajala** [comment]
Testing this with the latest master will be blocked on #1869

**Comment by elliottslaughter** [comment]
That was unintentional, I'm trying to assign bugs to the 25.06 milestone but it seems to have unintended consequences.

**Comment by apryakhin** [comment]
@syamajala This has been fixed. Can you please confirm whether you still see the issue or confirm that you don't see it? And if you don't can please close it?

**Comment by syamajala** [comment]
I am seeing a different assertion now.

**Comment by syamajala** [comment]
I opened #1871

**Comment by syamajala** [comment]
Or is it the same assertion and things just got renamed?

**Comment by syamajala** [comment]
@apryakhin just so we document the history, the fix for this was to go back to the old barrier implementation right?
